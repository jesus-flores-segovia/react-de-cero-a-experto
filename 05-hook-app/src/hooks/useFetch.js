import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        return () => {
            isMounted.current = false;
            console.log("isMounted to false.");
        }
    }, [])

    useEffect(() => {
        console.log("useEffect of [url].");
        setState({data: null, loading: true, error: null});

        fetch(url)
            .then(res => res.json())
            .then(data => {
                
                    setTimeout(() => {
                        if(isMounted.current){
                            setState({
                                loading: false,
                                error: null,
                                data
                            });
                        } else{
                            console.log("setState not called.");
                        }
                    }, 4000);
            });

    }, [url]);

    return state;
}
