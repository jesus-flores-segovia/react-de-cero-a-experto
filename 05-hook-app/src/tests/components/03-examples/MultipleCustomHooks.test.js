import {act, renderHook} from "@testing-library/react-hooks";
import { shallow } from "enzyme";
import {MultipleCustomHooks} from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("Tests inside file MultipleCustomHooks.test.js", () => {

    test("'MultipleCustomHooks' component must show correctly", () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        useCounter.mockReturnValue({
            counter: 1,
            increment: jest.fn(),
            reset: jest.fn()
        });

        const wrapper = shallow(<MultipleCustomHooks/>)
        
        expect(wrapper).toMatchSnapshot();
    });

    test("'MultipleCustomHooks' component must show the retrieved information", () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: "Jesús",
                quote: "This is an example quote."
            }],
            loading: false,
            error: null
        });

        useCounter.mockReturnValue({
            counter: 1,
            increment: jest.fn(),
            reset: jest.fn()
        });

        const wrapper = shallow(<MultipleCustomHooks/>)

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert").exists()).toBe(false);
        expect(wrapper.find(".mb-3").text().trim()).toBe("This is an example quote.");
        expect(wrapper.find("footer").text().trim()).toBe("Jesús");


    });
});