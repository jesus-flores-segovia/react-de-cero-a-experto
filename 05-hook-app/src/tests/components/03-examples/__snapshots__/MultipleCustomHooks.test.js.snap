// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file MultipleCustomHooks.test.js 'MultipleCustomHooks' component must show correctly 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert, alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;

exports[`Tests inside file MultipleCustomHooks.test.js 'MultipleCustomHooks' component must show the retrieved information 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-end"
  >
    <p
      className="mb-3"
    >
      This is an example quote.
    </p>
    <footer
      className="blockquote-footer"
    >
      Jes√∫s
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
